File: algo/valid-parentheses/isValid.go
func isValid(s string) bool {
    pmap := map[rune]rune{'[' : ']', '(' : ')', '{' : '}'}
    stack := []rune{}

    for _, r := range s {
        if v, ok := pmap[r]; ok {
            stack = append(stack, v)
        } else {
            if len(stack) == 0 {
                return false
            } else {
                last := len(stack) - 1
                if stack[last] == r {
                    stack = stack[:last]
                } else {
                    return false
                }
            }
        }
    }    
    return len(stack) == 0
}

File: algo/valid-parentheses/isValid.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        pmap = {'{' : '}', '(' : ')', '[' : ']'}
        
        for c in s:
            if c in pmap:
                stack.append(pmap[c])
            else:
                if len(stack) == 0:
                    return False
                else:
                    if stack[-1] != c:
                        return False
                    else:
                        stack = stack[:-1]
        return len(stack) == 0